### PR TITLE
make job definitions editable

### LINKED ISSUE #87: Enable Job Definitions to be edited
### Problem

In our initial work on adding job scheduling, we probably won't get to making Job Definitions editable. This issue is to track that follow on work. 

### Proposed Solution

The list and detail view for a Job Definition should have an Edit action that allows the user to reconfigure elements of the scheduled job. See the REST API for which fields are editable:

https://github.com/jupyter-server/jupyter-scheduler/pull/45

### LINKED ISSUE #172: Repurpose create form as "edit" form for job definitions
### Problem

Users want to be able to edit a job definition (JD) and either save it in place, or create a new job definition (aka "cloning" a job definition). See #87.

### Proposed Solution

Add a button called "Edit Job Definition" (to be confirmed) on the Job Definition Detail page. When the user clicks this button, they are brought to the Create Job form with the job definition detail prepopulated.

The Create Job form:
1. May benefit from having a different title ("Edit _[job definition name]_", for example)
2. Should have buttons "Cancel," "Save Changes," and "Save as New Job Definition"
  a. **Cancel** returns the user to the Job Definition Detail page from which they came (it should _not_ send the user to the JD list page, as it does today)
  b. **Save Changes** should, if all inputs are valid and if the API request succeeds, return users to the JD detail page from which they came.
  c. **Save as New Job Definition** should, if all inputs are valid and if the API request succeeds, take users to the JD list page, as the "Create JD" form does today.
4. Might _not_ let users switch to "Run now" — to be discussed

### Additional context

If the schedule of the job definition was created using an "easy scheduling" option, we should prepopulate the create form so that easy scheduling appears by default, with options preselected. (This may require you to interpret the cron schedule.)

### PR DESCRIPTION
**Pull request by dlqqq**
## Demo

https://user-images.githubusercontent.com/44106031/199151075-80af9de8-37b1-446b-abee-5522ba4af25a.mov

## Description

- Refactors `CreateScheduleOptions` and `ScheduleInputs` components to not require parents to pass handlers
  - These components are already able to change model state, so handling input changes should be abstracted by the components themselves.
  - Allows for separate `EditJobDefinitionForm` component to use `CreateScheduleOptions` without duplicating handler code
- Adds `EditJobDefinition` mainview component, rendered by setting `jobsView = JobsView.EditJobDefinition`
  - allows users to edit job definitions, as the name implies.

## Not included

- "Save as new job definition"
  - Not currently possible because we no longer save the input URI
- "populate easy scheduling"
  - Implementation would likely have to rely on complex regexes, and would be brittle. My suggestion is to make backend changes to allow for "schedule formats", which would allow schedules to be specified as something other than purely CRON expressions. This way, the backend could persist the initial schedule format the user specified, and save the UI the hassle of translating a CRON expression back into an easily-editable format.

## Related issues

- Closes #87 
- Closes #172

### CODE REVIEW

**Comment by dlqqq** [comment]
@jweill-aws Please verify functionality of schedule input handling & validation.

File: src/components/schedule-inputs.tsx [hunk 1]
```diff
@@ -252,3 +489,97 @@ export function ScheduleInputs(props: ScheduleInputsProps): JSX.Element | null {
...
+  }
+
+  return [hours, minutes];
+}
+
+/**
+ * Accepts a translation bundle in its constructor and returns the appropriate
+ * error message. If no errors are present, the methods return an empty string.
+ */
+export class ScheduleValidator {
```
Comment (jweill-aws) [inline src/components/schedule-inputs.tsx:513]: This file is getting a little sprawling; should `ScheduleValidator` be moved to a source file of its own?

File: src/components/schedule-inputs.tsx [hunk 2]
```diff
@@ -57,6 +52,248 @@ export function ScheduleInputs(props: ScheduleInputsProps): JSX.Element | null {
...
+
+    let minutes = props.model.scheduleHourMinute;
+    let schedule = props.model.schedule;
+
+    const scheduleHourMinuteError = validator.validateHourMinute(value);
+
+    if (!scheduleHourMinuteError) {
+      minutes = parseInt(value);
+      // No errors; compose a new schedule in cron format
+      switch (props.model.scheduleInterval) {
```
Comment (jweill-aws) [inline src/components/schedule-inputs.tsx:163]: This `switch` pattern works in other functions in this file, but in cases like this where there is only one case, I think it's safe to assume that the one case is true. Best to remove the `switch` block and assume `…scheduleInterval === 'hour'`.

File: src/components/schedule-inputs.tsx [hunk 3]
```diff
@@ -57,6 +52,248 @@ export function ScheduleInputs(props: ScheduleInputsProps): JSX.Element | null {
...
+
+    let monthDay = props.model.scheduleMonthDay;
+    let schedule = props.model.schedule;
+
+    const monthDayError = validator.validateMonthDay(value);
+
+    if (!monthDayError) {
+      monthDay = parseInt(value);
+      // Compose a new schedule in cron format
+      switch (props.model.scheduleInterval) {
```
Comment (jweill-aws) [inline src/components/schedule-inputs.tsx:196]: One-case `switch` block not needed now

**Review by jweill-aws — APPROVED** [review]
Generally looking good; just a few suggestions for cleanup

File: src/components/schedule-inputs.tsx (hunk 1 above)
Comment (dlqqq) [reply src/components/schedule-inputs.tsx:513]: If it's not being used anywhere else, I think it's best to keep it local to the file it's being used in. IMO, traversing long import paths is worse than having big files in terms of readability and maintainability. Open to feedback on this.

File: src/components/schedule-inputs.tsx (hunk 2 above)
Comment (dlqqq) [reply src/components/schedule-inputs.tsx]: Done. :+1:

File: src/components/schedule-inputs.tsx (hunk 3 above)
Comment (dlqqq) [reply src/components/schedule-inputs.tsx]: Done. :+1:

**Comment by 3coins** [comment]
Tested this last night, editing the schedule seems to be working as expected.

**Comment by dlqqq** [comment]
Let's hold off on merging until we get consensus on the UX. Let me rebase in the meantime.

**Review by andrii-i — APPROVED** [review]
Great job. Left some nit-picky (Nit) comments on type inference and interface conventions

File: src/mainviews/edit-job-definition.tsx [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        idPrefix=""
+        model={props.model}
+        handleModelChange={props.handleModelChange}
+        errors={fieldErrors}
+        handleErrorsChange={newErrors => setFieldErrors(newErrors)}
+      />
+      <Cluster gap={3} justifyContent="flex-end">
+        {loading ? (
+          <>
+            {trans.__('Saving changes...')}
```
Comment (jweill-aws) [inline src/mainviews/edit-job-definition.tsx:66]:
```suggestion
            {trans.__('Saving changes …')}
```

Minor: Replace with "space ellipsis" for consistency with the style used in the rest of this feature

File: src/mainviews/edit-job-definition.tsx [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+            >
+              {trans.__('Cancel')}
+            </Button>
+            <Button
+              color="primary"
+              variant="contained"
+              onClick={handleSubmit}
+              disabled={hasErrors}
+            >
+              {trans.__('Save changes')}
```
Comment (jweill-aws) [inline src/mainviews/edit-job-definition.tsx:85]:
```suggestion
              {trans.__('Save Changes')}
```

Minor: Use Title Case for button labels

**Comment by jweill-aws** [comment]
I tried to change the schedule to "06:05 on the 15th of the month" but the hour and minute did not propagate in the schedule. See video.

https://user-images.githubusercontent.com/93281816/199319082-13ee9479-da26-4723-933b-5d47448f435a.mov

File: src/mainviews/edit-job-definition.tsx [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+  const trans = useTranslator('jupyterlab');
+
+  return (
+    <Box sx={{ p: 4 }}>
+      <Stack spacing={4}>
+        <Breadcrumbs aria-label="breadcrumb">
+          <Link
+            underline="hover"
+            color="inherit"
+            onClick={(): void => {
```
[the remark this answers is not in the document]
Comment (dlqqq) [reply src/mainviews/edit-job-definition.tsx:104]: Happy to remove return type, but I prefer not giving references to arguments I don't use.

[the remark this answers is not in the document]
Comment (andrii-i) [reply src/mainviews/edit-job-definition.tsx:104]: Understandable, sounds good

**Comment by andrii-i** [comment]
> I tried to change the schedule to "06:05 on the 15th of the month" but the hour and minute did not propagate in the schedule. See video.

Reproduced the bug

**Comment by dlqqq** [comment]
Bug fixed by this commit: https://github.com/jupyter-server/jupyter-scheduler/pull/238/commits/1d9b870c7cb9794340e652d6c142b4ac7c0a3dd3

This is why duplicate state (between `model.schedule` and all the other input fields) is evil. Forgetting to update one of them leads to weird errors like this.

**Comment by jweill-aws** [comment]
Found another bug: 

1. Schedule is 0 0 15 * *  (00:00 on the 15th of the month)
2. Click "Edit job definition"
3. Drop down "Interval," pick "Month"
4. **Error:** "Day of the month" is blank (should be "15")

**Review by jweill-aws — APPROVED** [review]

**Pull request merged**